File: android/src/main/kotlin/com/technit/capacitor/conekta/ConektaTokenizer.kt
package com.technit.capacitor.conekta

import android.annotation.SuppressLint
import android.app.Activity
import android.graphics.Bitmap
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.webkit.JavascriptInterface
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import org.json.JSONObject

class ConektaTokenizer {
    companion object {
        private const val TAG = "ConektaTokenizer"
        const val SDK_READY_TIMEOUT_MS: Long = 15_000
        const val TOKEN_REQUEST_TIMEOUT_MS: Long = 20_000

        private val HTML = """
            <!DOCTYPE html>
            <html>
            <head>
            <script src="https://cdn.conekta.io/js/latest/conekta.js"></script>
            <script>
            window.addEventListener('load', function() {
                ConektaBridge.onResult(JSON.stringify({type:'ready'}));
            });
            function initConekta(publicKey) {
                try {
                    if (typeof Conekta === 'undefined' || !Conekta) {
                        ConektaBridge.onResult(JSON.stringify({
                            type:'token', success:false,
                            error: { message_to_purchaser: 'Conekta SDK not loaded', code: 'sdk_not_loaded' }
                        }));
                        return;
                    }
                    Conekta.setPublicKey(publicKey);
                    Conekta.setLanguage('es');
                } catch (e) {
                    ConektaBridge.onResult(JSON.stringify({
                        type:'token', success:false,
                        error: { message_to_purchaser: (e && e.message) || 'initConekta threw', code: 'js_exception' }
                    }));
                }
            }
            function createToken(name, number, cvc, expMonth, expYear) {
                try {
                    if (typeof Conekta === 'undefined' || !Conekta || !Conekta.Token) {
                        ConektaBridge.onResult(JSON.stringify({
                            type:'token', success:false,
                            error: { message_to_purchaser: 'Conekta SDK not loaded', code: 'sdk_not_loaded' }
                        }));
                        return;
                    }
                    Conekta.Token.create({
                        card: { name: name, number: number, cvc: cvc, exp_month: expMonth, exp_year: expYear }
                    }, function(token) {
                        ConektaBridge.onResult(JSON.stringify({type:'token', success:true, token:token}));
                    }, function(error) {
                        ConektaBridge.onResult(JSON.stringify({type:'token', success:false, error: error || {message_to_purchaser: 'Token creation failed'}}));
                    });
                } catch (e) {
                    ConektaBridge.onResult(JSON.stringify({
                        type:'token', success:false,
                        error: { message_to_purchaser: (e && e.message) || 'createToken threw', code: 'js_exception' }
                    }));
                }
            }
            </script>
            </head>
            <body></body>
            </html>
        """.trimIndent()
    }

    data class ConektaFailure(val raw: JSONObject, val code: String?)

    private var webView: WebView? = null
    private var publicKey: String? = null
    private var sdkReady = false
    private var sdkReadyWaiters: MutableList<(Boolean, ConektaFailure?) -> Unit> = mutableListOf()
    private var tokenSuccess: ((JSONObject) -> Unit)? = null
    private var tokenError: ((ConektaFailure) -> Unit)? = null
    private var sdkReadyTimeout: Runnable? = null
    private var tokenTimeout: Runnable? = null
    private var hasAttemptedReload = false
    private lateinit var mainHandler: Handler

    @SuppressLint("SetJavaScriptEnabled")
    fun setup(activity: Activity) {
        mainHandler = Handler(Looper.getMainLooper())
        mainHandler.post {
            if (webView != null) return@post
            val wv = WebView(activity)
            wv.settings.javaScriptEnabled = true
            wv.settings.domStorageEnabled = true
            wv.addJavascriptInterface(this, "ConektaBridge")
            wv.webViewClient = object : WebViewClient() {
                override fun onReceivedError(
                    view: WebView,
                    request: WebResourceRequest,
                    error: WebResourceError
                ) {
                    if (request.isForMainFrame) {
                        handleNavigationFailure()
                    }
                }
            }
            wv.loadDataWithBaseURL("https://conekta.com", HTML, "text/html", "UTF-8", null)
            webView = wv
        }
    }

    fun setPublicKey(publicKey: String) {
        this.publicKey = publicKey
    }

    fun isReady(): Boolean = sdkReady

    fun warmUp(
        activity: Activity,
        onSuccess: () -> Unit,
        onError: (ConektaFailure) -> Unit
    ) {
        if (!::mainHandler.isInitialized) setup(activity)
        ensureReady { ok, failure ->
            if (ok) onSuccess() else onError(failure ?: sdkTimeoutFailure())
        }
    }

    private fun ensureReady(callback: (Boolean, ConektaFailure?) -> Unit) {
        if (sdkReady) {
            callback(true, null)
            return
        }
        sdkReadyWaiters.add(callback)
        if (sdkReadyTimeout == null) {
            val r = Runnable { flushReadyWaiters(false, sdkTimeoutFailure()) }
            sdkReadyTimeout = r
            mainHandler.postDelayed(r, SDK_READY_TIMEOUT_MS)
        }
    }

    private fun flushReadyWaiters(ok: Boolean, failure: ConektaFailure?) {
        sdkReadyTimeout?.let { mainHandler.removeCallbacks(it) }
        sdkReadyTimeout = null
        val waiters = sdkReadyWaiters.toList()
        sdkReadyWaiters.clear()
        for (w in waiters) w(ok, failure)
    }

    private fun sdkTimeoutFailure(): ConektaFailure {
        val raw = JSONObject()
            .put("message_to_purchaser", "Conekta SDK did not load in time.")
            .put("code", "sdk_load_timeout")
        return ConektaFailure(raw, "sdk_load_timeout")
    }

    private fun tokenTimeoutFailure(): ConektaFailure {
        val raw = JSONObject()
            .put("message_to_purchaser", "Conekta token request timed out.")
            .put("code", "token_request_timeout")
        return ConektaFailure(raw, "token_request_timeout")
    }

    fun createToken(
        name: String,
        cardNumber: String,
        expMonth: String,
        expYear: String,
        cvc: String,
        onSuccess: (JSONObject) -> Unit,
        onError: (ConektaFailure) -> Unit
    ) {
        val key = publicKey
        if (key == null) {
            val raw = JSONObject()
                .put("message_to_purchaser", "Public key not set. Call setPublicKey() before createToken().")
                .put("code", "public_key_not_set")
            onError(ConektaFailure(raw, "public_key_not_set"))
            return
        }

        if (tokenSuccess != null || tokenError != null) {
            val raw = JSONObject()
                .put("message_to_purchaser", "A Conekta token request is already in flight.")
                .put("code", "request_in_flight")
            onError(ConektaFailure(raw, "request_in_flight"))
            return
        }

        tokenSuccess = onSuccess
        tokenError = onError

        ensureReady { ok, failure ->
            if (!ok) {
                finishTokenError(failure ?: sdkTimeoutFailure())
                return@ensureReady
            }
            mainHandler.post {
                val wv = webView
                if (wv == null) {
                    val raw = JSONObject()
                        .put("message_to_purchaser", "Conekta WebView is not ready.")
                        .put("code", "webview_not_ready")
                    finishTokenError(ConektaFailure(raw, "webview_not_ready"))
                    return@post
                }

                val escapedKey = key.replace("'", "\\'")
                val escapedName = name.replace("'", "\\'")

                armTokenTimeout()

                wv.evaluateJavascript("initConekta('$escapedKey')", null)
                wv.evaluateJavascript(
                    "createToken('$escapedName', '$cardNumber', '$cvc', '$expMonth', '$expYear')",
                    null
                )
            }
        }
    }

    private fun armTokenTimeout() {
        tokenTimeout?.let { mainHandler.removeCallbacks(it) }
        val r = Runnable { finishTokenError(tokenTimeoutFailure()) }
        tokenTimeout = r
        mainHandler.postDelayed(r, TOKEN_REQUEST_TIMEOUT_MS)
    }

    private fun finishTokenSuccess(tokenObj: JSONObject) {
        tokenTimeout?.let { mainHandler.removeCallbacks(it) }
        tokenTimeout = null
        val success = tokenSuccess
        tokenSuccess = null
        tokenError = null
        success?.invoke(tokenObj)
    }

    private fun finishTokenError(failure: ConektaFailure) {
        tokenTimeout?.let { mainHandler.removeCallbacks(it) }
        tokenTimeout = null
        val err = tokenError
        tokenSuccess = null
        tokenError = null
        err?.invoke(failure)
    }

    private fun handleNavigationFailure() {
        if (!hasAttemptedReload) {
            hasAttemptedReload = true
            webView?.loadDataWithBaseURL("https://conekta.com", HTML, "text/html", "UTF-8", null)
            return
        }
        val raw = JSONObject()
            .put("message_to_purchaser", "Conekta WebView is not ready.")
            .put("code", "webview_not_ready")
        flushReadyWaiters(false, ConektaFailure(raw, "webview_not_ready"))
        finishTokenError(ConektaFailure(raw, "webview_not_ready"))
    }

    @JavascriptInterface
    fun onResult(jsonStr: String) {
        mainHandler.post {
            try {
                val json = JSONObject(jsonStr)
                when (json.optString("type")) {
                    "ready" -> {
                        sdkReady = true
                        flushReadyWaiters(true, null)
                    }
                    "token" -> {
                        val success = json.optBoolean("success", false)
                        if (success) {
                            val tokenObj = json.optJSONObject("token") ?: JSONObject()
                            finishTokenSuccess(tokenObj)
                        } else {
                            val raw = json.optJSONObject("error")
                                ?: JSONObject().put(
                                    "message_to_purchaser",
                                    json.optString("error", "Token creation failed")
                                )
                            val code = if (raw.has("code")) raw.optString("code") else null
                            finishTokenError(ConektaFailure(raw, code))
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error parsing WebView result", e)
                val raw = JSONObject().put("message_to_purchaser", e.message ?: "Unknown error")
                finishTokenError(ConektaFailure(raw, null))
            }
        }
    }
}

File: ios/Sources/ConektaTokenizerPlugin/ConektaTokenizer.swift
import Foundation
import WebKit

public class ConektaTokenizer: NSObject, WKScriptMessageHandler, WKNavigationDelegate {
    public static let sdkReadyTimeout: TimeInterval = 15
    public static let tokenRequestTimeout: TimeInterval = 20

    private var webView: WKWebView?
    private var publicKey: String?
    private var sdkReady = false
    private var sdkReadyWaiters: [(Result<Void, Error>) -> Void] = []
    private var tokenCompletion: ((Result<[String: Any], Error>) -> Void)?
    private var sdkReadyTimeoutItem: DispatchWorkItem?
    private var tokenTimeoutItem: DispatchWorkItem?
    private var hasAttemptedReload = false

    private static let html = """
    <!DOCTYPE html>
    <html>
    <head>
    <script src="https://cdn.conekta.io/js/latest/conekta.js"></script>
    <script>
    window.addEventListener('load', function() {
        window.webkit.messageHandlers.conektaResult.postMessage(JSON.stringify({type:'ready'}));
    });
    function initConekta(publicKey) {
        try {
            if (typeof Conekta === 'undefined' || !Conekta) {
                window.webkit.messageHandlers.conektaResult.postMessage(JSON.stringify({
                    type:'token', success:false,
                    error: { message_to_purchaser: 'Conekta SDK not loaded', code: 'sdk_not_loaded' }
                }));
                return;
            }
            Conekta.setPublicKey(publicKey);
            Conekta.setLanguage('es');
        } catch (e) {
            window.webkit.messageHandlers.conektaResult.postMessage(JSON.stringify({
                type:'token', success:false,
                error: { message_to_purchaser: (e && e.message) || 'initConekta threw', code: 'js_exception' }
            }));
        }
    }
    function createToken(name, number, cvc, expMonth, expYear) {
        try {
            if (typeof Conekta === 'undefined' || !Conekta || !Conekta.Token) {
                window.webkit.messageHandlers.conektaResult.postMessage(JSON.stringify({
                    type:'token', success:false,
                    error: { message_to_purchaser: 'Conekta SDK not loaded', code: 'sdk_not_loaded' }
                }));
                return;
            }
            Conekta.Token.create({
                card: { name: name, number: number, cvc: cvc, exp_month: expMonth, exp_year: expYear }
            }, function(token) {
                window.webkit.messageHandlers.conektaResult.postMessage(JSON.stringify({type:'token', success:true, token:token}));
            }, function(error) {
                window.webkit.messageHandlers.conektaResult.postMessage(JSON.stringify({type:'token', success:false, error: error || {message_to_purchaser: 'Token creation failed'}}));
            });
        } catch (e) {
            window.webkit.messageHandlers.conektaResult.postMessage(JSON.stringify({
                type:'token', success:false,
                error: { message_to_purchaser: (e && e.message) || 'createToken threw', code: 'js_exception' }
            }));
        }
    }
    </script>
    </head>
    <body></body>
    </html>
    """

    public func setup() {
        DispatchQueue.main.async {
            guard self.webView == nil else { return }
            let config = WKWebViewConfiguration()
            config.userContentController.add(self, name: "conektaResult")
            let wv = WKWebView(frame: .zero, configuration: config)
            wv.navigationDelegate = self
            self.webView = wv
            wv.loadHTMLString(ConektaTokenizer.html, baseURL: URL(string: "https://conekta.com"))
        }
    }

    public func setPublicKey(_ publicKey: String) {
        self.publicKey = publicKey
    }

    public func isReady() -> Bool {
        return sdkReady
    }

    public func warmUp(completion: @escaping (Result<Void, Error>) -> Void) {
        ensureReady(completion: completion)
    }

    private func ensureReady(completion: @escaping (Result<Void, Error>) -> Void) {
        if sdkReady {
            completion(.success(()))
            return
        }
        sdkReadyWaiters.append(completion)
        if webView == nil {
            setup()
        }
        if sdkReadyTimeoutItem == nil {
            let item = DispatchWorkItem { [weak self] in
                self?.flushReadyWaiters(with: .failure(ConektaError.sdkLoadTimeout))
            }
            sdkReadyTimeoutItem = item
            DispatchQueue.main.asyncAfter(deadline: .now() + ConektaTokenizer.sdkReadyTimeout, execute: item)
        }
    }

    private func flushReadyWaiters(with result: Result<Void, Error>) {
        sdkReadyTimeoutItem?.cancel()
        sdkReadyTimeoutItem = nil
        let waiters = sdkReadyWaiters
        sdkReadyWaiters.removeAll()
        for waiter in waiters {
            waiter(result)
        }
    }

    public func createToken(
        name: String,
        cardNumber: String,
        expMonth: String,
        expYear: String,
        cvc: String,
        completion: @escaping (Result<[String: Any], Error>) -> Void
    ) {
        guard let publicKey = self.publicKey else {
            completion(.failure(ConektaError.publicKeyNotSet))
            return
        }

        if tokenCompletion != nil {
            completion(.failure(ConektaError.requestInFlight))
            return
        }

        self.tokenCompletion = completion

        ensureReady { [weak self] readyResult in
            guard let self = self else { return }
            if case .failure(let error) = readyResult {
                self.finishToken(with: .failure(error))
                return
            }
            guard let webView = self.webView else {
                self.finishToken(with: .failure(ConektaError.webViewNotReady))
                return
            }

            DispatchQueue.main.async {
                let initJS = "initConekta('\(publicKey.replacingOccurrences(of: "'", with: "\\'"))'); 0"
                let tokenJS = "createToken('\(name.replacingOccurrences(of: "'", with: "\\'"))', '\(cardNumber)', '\(cvc)', '\(expMonth)', '\(expYear)'); 0"

                self.armTokenTimeout()

                webView.evaluateJavaScript(initJS) { [weak self] _, error in
                    guard let self = self else { return }
                    if let error = error {
                        self.finishToken(with: .failure(error))
                        return
                    }
                    webView.evaluateJavaScript(tokenJS) { [weak self] _, error in
                        guard let self = self else { return }
                        if let error = error {
                            self.finishToken(with: .failure(error))
                        }
                    }
                }
            }
        }
    }

    private func armTokenTimeout() {
        tokenTimeoutItem?.cancel()
        let item = DispatchWorkItem { [weak self] in
            self?.finishToken(with: .failure(ConektaError.tokenRequestTimeout))
        }
        tokenTimeoutItem = item
        DispatchQueue.main.asyncAfter(deadline: .now() + ConektaTokenizer.tokenRequestTimeout, execute: item)
    }

    private func finishToken(with result: Result<[String: Any], Error>) {
        tokenTimeoutItem?.cancel()
        tokenTimeoutItem = nil
        let completion = tokenCompletion
        tokenCompletion = nil
        completion?(result)
    }

    // MARK: - WKScriptMessageHandler

    public func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard let body = message.body as? String,
              let data = body.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else {
            return
        }

        switch type {
        case "ready":
            sdkReady = true
            flushReadyWaiters(with: .success(()))
        case "token":
            let success = json["success"] as? Bool ?? false
            if success, let tokenObj = json["token"] as? [String: Any] {
                finishToken(with: .success(tokenObj))
            } else {
                let raw: [String: Any]
                if let dict = json["error"] as? [String: Any] {
                    raw = dict
                } else if let str = json["error"] as? String {
                    raw = ["message_to_purchaser": str]
                } else {
                    raw = [:]
                }
                let message = (raw["message_to_purchaser"] as? String)
                    ?? (raw["message"] as? String)
                    ?? "Token creation failed"
                finishToken(with: .failure(ConektaError.apiError(
                    message: message,
                    code: raw["code"] as? String,
                    type: raw["type"] as? String,
                    param: raw["param"] as? String,
                    raw: raw
                )))
            }
        default:
            break
        }
    }

    // MARK: - WKNavigationDelegate

    public func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        handleNavigationFailure(error: error)
    }

    public func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        handleNavigationFailure(error: error)
    }

    private func handleNavigationFailure(error: Error) {
        if !hasAttemptedReload, let wv = webView {
            hasAttemptedReload = true
            wv.loadHTMLString(ConektaTokenizer.html, baseURL: URL(string: "https://conekta.com"))
            return
        }
        flushReadyWaiters(with: .failure(ConektaError.webViewNotReady))
        finishToken(with: .failure(ConektaError.webViewNotReady))
    }
}

enum ConektaError: LocalizedError {
    case publicKeyNotSet
    case webViewNotReady
    case sdkLoadTimeout
    case tokenRequestTimeout
    case requestInFlight
    case apiError(message: String, code: String?, type: String?, param: String?, raw: [String: Any])

    var errorDescription: String? {
        switch self {
        case .publicKeyNotSet:
            return "Public key not set. Call setPublicKey() before createToken()."
        case .webViewNotReady:
            return "Conekta WebView is not ready."
        case .sdkLoadTimeout:
            return "Conekta SDK did not load in time."
        case .tokenRequestTimeout:
            return "Conekta token request timed out."
        case .requestInFlight:
            return "A Conekta token request is already in flight."
        case .apiError(let message, _, _, _, _):
            return message
        }
    }

    var code: String? {
        switch self {
        case .publicKeyNotSet: return "public_key_not_set"
        case .webViewNotReady: return "webview_not_ready"
        case .sdkLoadTimeout: return "sdk_load_timeout"
        case .tokenRequestTimeout: return "token_request_timeout"
        case .requestInFlight: return "request_in_flight"
        case .apiError(_, let code, _, _, _): return code
        }
    }
}
